### PR TITLE
[SIG-43614] Fix SQL parsing for select statement as function argument

### PR DESCRIPTION
**Pull request by sinanunan**
we were failing to parse `SELECT parse_json(SELECT column2 FROM values(1, 'null'))` even though it is valid sql. This was happening because the parser expects every `SELECT` subquery to be enclosed by parentheses (hence, in a function call, by 2). This assumption is incorrect, and even the test added to confirm this is only a false positive. For example, `SELECT SELECT column2` actually succeeds. 

This PR disables the `parse_invalid_subquery_without_parens`, fixes the sql parsing issue and adds a test for it.

Confirmed on local qwill that the sql does not fail anymore.
```
WITH 
  venue_list as ( 
    select 
      TO_CHAR ( 
        ARRAY_CONSTRUCT_COMPACT ( 
          {{ venue_one }}, 
          {{ venue_two }}, 
          {{ venue_three }}, 
          {{ venue_four }}, 
          {{ venue_five }} 
        ) 
      ) as property 
  ), 
  query_one AS ( 
    select distinct 
      value as property 
    from 
      TABLE ( 
        FLATTEN ( 
          input => parse_json ( 
            select 
              property 
            from 
              venue_list 
          ) 
        ) 
      ) 
  ) 
SELECT 
  qo.property::string AS row_venue, 
  qt.property::string AS column_venue 
FROM 
  query_one AS qo 
  CROSS JOIN query_one AS qt

```

### CODE REVIEW

File: tests/sqlparser_common.rs [hunk 1]
```diff
@@ -2975,6 +2975,9 @@ fn parse_drop_view() {
 }
 
 #[test]
+#[ignore]
+// XXX (sinan) This test was passing even though the parser sometimes incorrectly parses invalid subqueries (such as SELECT SELECT column) as
+// valid sql. It started to fail after the introduction of parse_select_as_func_argument. Disabling the test for now.
```
Comment (sigmacomputingluke) [inline tests/sqlparser_common.rs:2980]: Not sure I understand. Is the test bogus? Is it not getting us the desired coverage? Both?

File: src/parser.rs [hunk 2]
```diff
@@ -1221,8 +1226,7 @@ impl<'a> Parser<'a> {
 
     /// Return the first unprocessed token, possibly whitespace.
     pub fn next_token_no_skip(&mut self) -> Option<&Token> {
-        self.index += 1;
-        self.tokens.get(self.index - 1)
+        self.tokens.get(self.index)
```
Comment (sigmacomputingluke) [inline src/parser.rs:1229]: Why this change?

Comment (sinanunan) [reply src/parser.rs:1229]: Because it was skipping to the next token.

File: tests/sqlparser_common.rs (hunk 1 above)
Comment (jmhain) [reply tests/sqlparser_common.rs:2980]: This test failing is an indicator that this isn't the right solution. It looks like this PR is trying to always allow subqueries without parens, whereas what we need to support is the Snowflake quirk that subqueries can be passed in unary function calls without an extra set of parens.

Comment (sinanunan) [reply tests/sqlparser_common.rs]: @sigmacomputingluke @jmhain Thanks for the explanation. I've verified that subqueries without parentheses are allowed in functions without named arguments, and added tests for it. I've also reenabled the test I disabled.

Please let me know if any other cases pops up, I'll add them to the tests

File: tests/sqlparser_snowflake.rs [hunk 3]
```diff
@@ -151,6 +151,43 @@ fn test_single_table_in_parenthesis_with_alias() {
...
+
+        // named function arguments should not work
+        assert_eq!(snowflake().parse_sql_statements(
+            "SELECT func(expr => SELECT 1)",
+           
+        ).unwrap_err(),
+        ParserError::ParserError(
+            "SELECT func(expr => SELECT".to_string(),
+            "Expected ), found: 1".to_string()
+        ));
```
Comment (sigmacomputingluke) [inline tests/sqlparser_snowflake.rs:188]: Nit: indentation

File: src/parser.rs [hunk 4]
```diff
@@ -1221,8 +1221,7 @@ impl<'a> Parser<'a> {
...
```
Comment (sigmacomputingluke) [inline src/parser.rs:1224]: Is this change still needed with the new approach?

**Review by sigmacomputingluke — APPROVED** [review]
LGTM!

File: src/parser.rs (hunk 4 above)
Comment (sinanunan) [reply src/parser.rs:1224]: it was never needed. The function only used in parsing tab separated values, but it is also very useful for debugging.  Don't see any reason not to include it

Comment (sigmacomputingluke) [reply src/parser.rs:1224]: I see. I assumed there was some reason for the old code being written that way but I guess it may just be historical accident. We have test coverage that would catch if this change broke the tab separated values case, right?

Comment (sinanunan) [reply src/parser.rs]: At first look, it looks like the change will not break the functionality, but there is not enough test coverage. I'm adding a comment on the code and reverting the fix
